sdk/java-sdk-protobuf/src/main/scala/kalix/javasdk/impl/telemetry/Telemetry.scala

```diff
@@ -0,0 +1,153 @@
+import io.opentelemetry.sdk.trace.`export`.SimpleSpanProcessor
+import io.opentelemetry.semconv.resource.attributes.ResourceAttributes
+import kalix.javasdk.Metadata
+import kalix.javasdk.impl.MetadataImpl
+import kalix.javasdk.impl.Service
+import kalix.protocol.action.ActionCommand
+import kalix.protocol.entity.Command
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+import scala.jdk.OptionConverters._
+
+object Telemetry {
+
+  val TRACE_PARENT_KEY = "traceparent"
+
+  val logger: Logger = LoggerFactory.getLogger(getClass)
+
+  val otelGetter = new TextMapGetter[Metadata]() {
+    override def get(carrier: Metadata, key: String): String = {
+      logger.debug("For the key [{}] the value is [{}]", key, carrier.get(key))
+      carrier.get(key).toScala.getOrElse("")
+    }
+
+    override def keys(carrier: Metadata): java.lang.Iterable[String] =
+      carrier.getAllKeys
+  }
+}
+
+class Telemetry(serviceName: String) {
+
+  import kalix.javasdk.impl.telemetry.Telemetry.logger
+
+  val openTelemetry: OpenTelemetry = {
+
+    val resource =
+      Resource.getDefault.merge(Resource.create(Attributes.of(ResourceAttributes.SERVICE_NAME, serviceName)))
+    val sdkTracerProvider =
+      SdkTracerProvider
+        .builder()
+        .addSpanProcessor(
+          SimpleSpanProcessor.create(OtlpGrpcSpanExporter.builder().setEndpoint("http://jaeger:4317").build()))
+        .setResource(resource)
+        .build()
+    val sdk = OpenTelemetrySdk
+      .builder()
+      .setTracerProvider(sdkTracerProvider)
+      .setPropagators(ContextPropagators.create(W3CTraceContextPropagator.getInstance()))
+      .build()
+    //    Runtime.getRuntime().addShutdownHook(new Thread(sdkTracerProvider.close)) TODO add shutdown
+    sdk
+  }
+
+  /**
+   * TODO
+   */
+  def buildSpan(service: Service, command: Command): Option[Span] = {
+    val metadata = new MetadataImpl(command.metadata.map(_.entries).getOrElse(Nil))
+    if (metadata.get(Telemetry.TRACE_PARENT_KEY).isPresent) {
+      logger.debug(" Service content [{}].", service)
+      logger.debug(" commandName content [{}].", command)
+
+      val context = openTelemetry.getPropagators.getTextMapPropagator
+        .extract(OtelContext.current(), metadata, Telemetry.otelGetter.asInstanceOf[TextMapGetter[Object]])
+      logger.debug("context from traceparent [{}].", context)
+      val tracer = openTelemetry.getTracer("java-sdk")
+      val span = tracer
+        .spanBuilder(s"""${service.serviceName}.${command.entityId}""")
+        .setParent(context)
+        .setSpanKind(SpanKind.SERVER)
+        .startSpan()
+      Some(
+        span
+          //          .setAttribute(SemanticAttributes.HTTP_METHOD, httpMethod.name)
+          .setAttribute("server.address", "localhost")
+          // TODO use Kubernetes .svc name?
+          .setAttribute("server.port", 8080) // TODO read the port from properties
+          .setAttribute("server.scheme", "http")
```

Comment:
Still not likely to be http, also, again what is the port, the public one that client used to hit the proxy or the internal one for the SDK?

If we don't have anything more correct/better here, maybe skip the attributes completely instead? The trace is anyway identified by the service name?

Reply:
I was going to checkout this attributes today, but got busy with the operator. Yeah, let's ditch these attributes by now and we can add them later. Once the spike gets some maturity.